File: day05/day5.py
import time
from itertools import cycle

import numpy as np


class Line:

    def __init__(self, start, end):
        self.start = np.array([start[0], start[1]])
        self.end = np.array([end[0], end[1]])

    def onGrid(self) -> bool:
        """Checks if line is vertical or horizontal

        :return: bool
        """
        if self.start[0] == self.end[0] or self.start[1] == self.end[1]:
            return True
        else:
            return False

    def __repr__(self):
        return f"{self.start} -> {self.end}"


def parseLineInput(lineInput: str) -> Line:
    startInput, endInput = lineInput.split(" -> ")
    startPoints = [int(x) for x in startInput.split(",")]
    endPoints = [int(x) for x in endInput.split(",")]
    line = Line(startPoints, endPoints)
    return line


def getBounds(lines: [Line]):
    points = []
    for line in lines:
        points.append(line.start)
        points.append(line.end)
    pointMatrix = np.stack(points)
    xMax = pointMatrix[:, 0].max()
    yMax = pointMatrix[:, 1].max()
    return xMax, yMax


def part1(lines):
    x, y = getBounds(lines)
    ground = np.zeros((x + 1, y + 1))

    for line in lines:
        if line.onGrid():
            xs, xe = line.start[0], line.end[0]
            ys, ye = line.start[1], line.end[1]
            xPoints = np.arange(xs, xe + 1) if xe > xs else np.arange(xe, xs + 1)
            yPoints = np.arange(ys, ye + 1) if ye > ys else np.arange(ye, ys + 1)
            for i in range(max(len(xPoints), len(yPoints))):
                x = xPoints[i % len(xPoints)]
                y = yPoints[i % len(yPoints)]
                ground[x, y] += 1

    return np.sum(ground >= 2)


def part2(lines):
    x, y = getBounds(lines)
    ground = np.zeros((x + 1, y + 1))

    for line in lines:
        xs, xe = line.start[0], line.end[0]
        ys, ye = line.start[1], line.end[1]
        xPoints = np.arange(xs, xe + 1) if xe > xs else np.arange(xs, xe - 1, -1)
        yPoints = np.arange(ys, ye + 1) if ye > ys else np.arange(ys, ye - 1, -1)
        for i in range(max(len(xPoints), len(yPoints))):
            x = xPoints[i % len(xPoints)]
            y = yPoints[i % len(yPoints)]
            ground[x, y] += 1

    return np.sum(ground >= 2)


if __name__ == '__main__':
    start = time.time()
    # parse input
    with open("input.txt") as f:
        lines = [parseLineInput(x) for x in f.readlines()]

    sol1 = part1(lines)
    sol2 = part2(lines)

    print(sol1)
    print(sol2)

    print(f"time: {(time.time() - start).__round__(6) * 1000 } ms")

File: day05/day5.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strconv"
	"strings"
	"time"
)

type Point struct {
	x, y int
}

type Line struct {
	start, end Point
}

func (line Line) isAxial() bool {
	if line.start.x == line.end.x || line.start.y == line.end.y {
		return true
	} else {
		return false
	}
}

func (line Line) Distance() int {
	xDistance := int(math.Abs(float64(line.end.x - line.start.x)))
	yDistance := int(math.Abs(float64(line.end.y - line.start.y)))

	if xDistance > yDistance {
		return xDistance
	} else {
		return yDistance
	}
}

func (line Line) Interpolate() []Point {
	xDir, yDir := 1, 1
	if line.end.x < line.start.x {
		xDir = -1
	} else if line.end.x == line.start.x {
		xDir = 0
	}

	if line.end.y < line.start.y {
		yDir = -1
	} else if line.end.y == line.start.y {
		yDir = 0
	}
	distance := line.Distance()
	points := make([]Point, distance+1)

	var x, y int

	for i := 0; i <= distance; i++ {
		x = line.start.x + i*xDir
		y = line.start.y + i*yDir
		point := Point{x, y}
		points[i] = point
	}
	return points
}

func parseLineInput(lineInput string) Line {
	inputs := strings.Split(lineInput, " -> ")
	startInputs := strings.Split(inputs[0], ",")
	endInputs := strings.Split(inputs[1], ",")

	startX, _ := strconv.Atoi(startInputs[0])
	startY, _ := strconv.Atoi(startInputs[1])
	endX, _ := strconv.Atoi(endInputs[0])
	endY, _ := strconv.Atoi(endInputs[1])

	start := Point{
		x: startX,
		y: startY,
	}

	end := Point{
		x: endX,
		y: endY,
	}

	return Line{
		start,
		end,
	}
}

func main() {
	start := time.Now()

	// Input reading
	bs, err := ioutil.ReadFile("input.txt")

	if err != nil {
		panic(err)
	}

	input := strings.Split(string(bs), "\n")

	sol1 := part1(input)
	sol2 := part2(input)

	end := time.Since(start)

	fmt.Println("Part 1: ", sol1)
	fmt.Println("Part 2: ", sol2)
	fmt.Println("Time: ", end)

}

func part1(input []string) int {
	var ground [1000][1000]int

	for _, lineInput := range input {
		line := parseLineInput(lineInput)
		if line.isAxial() {
			points := line.Interpolate()

			for _, point := range points {

				ground[point.x][point.y] += 1
			}
		}
	}

	var s int
	for _, row := range ground {
		for _, v := range row {
			if v >= 2 {
				s++
			}
		}
	}
	return s
}

func part2(input []string) int {
	var ground [1000][1000]int

	for _, lineInput := range input {
		line := parseLineInput(lineInput)
		points := line.Interpolate()

		for _, point := range points {

			ground[point.x][point.y] += 1
		}

	}

	var s int
	for _, row := range ground {
		for _, v := range row {
			if v >= 2 {
				s++
			}
		}
	}
	return s
}
